### DISCUSSION TITLE
Low performance issue on stopwatch

### DISCUSSION THREAD

**Discussion by Mark-Krensel** [neuefische/questions#284] [Web] [answered]
Hello, I created a stopwatch, which functionality is ok, but it seems to work kind of slow.
When the start button is clicked, it takes about 1s to show the actual stopped time, before it shows ```NaN```, but only when clicking it the first time.
After that, when the button is clicked it shows the local time for about 1s. If however it's clicked multiple times in fast succession, it shows for 1s the previously stopped time. 
Is this behaviour due to the time being tracked in 1s steps?

### Stopwatch

```jsx
import { useEffect, useState } from "react";

export default function Stopwatch() {
  const [startTime, setStartTime] = useState();
  const [hours, setHours] = useState(0);
  const [minutes, setMinutes] = useState(0);
  const [seconds, setSeconds] = useState(0);

  useEffect(() => {
    const interval = setInterval(() => {
      const now = new Date();
      const difference = now.getTime() - startTime;

      const h = Math.floor(
        (difference % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60)
      );
      setHours(h);

      const m = Math.floor((difference % (1000 * 60 * 60)) / (1000 * 60));
      setMinutes(m);

      const s = Math.floor((difference % (1000 * 60)) / 1000);
      setSeconds(s);
      localStorage.setItem("startTime", startTime);
    }, 1000);

    return () => clearInterval(interval);
  }, [startTime]);
  return (
    <>
      {startTime && (
        <>
          <div>
            {hours}:{minutes}:{seconds}
          </div>
          <button
            type="button"
            aria-label="stop stopwatch"
            onClick={() => {
              setStartTime("");
              localStorage.removeItem("startTime");
            }}
          >
            Stop
          </button>
        </>
      )}
      {!startTime && (
        <button
          type="button"
          aria-label="start stopwatch"
          onClick={() => {
            setStartTime(new Date().getTime());
          }}
        >
          Start
        </button>
      )}
    </>
  );
}

```

**Comment by ThomasF85** [comment, marked as answer]
Hi Mark,

your useEffect is running:
1. After the initial render of your component
2. Every time startTime changes

That it runs after the initial rendern (when startTime is undefined) is causing your Problem here. You have to only set your interval when startTime is not undefined.

**Reply by Mark-Krensel** [reply]
Thank you!

**Reply by Mark-Krensel** [reply]
So thanks to you first part of the problem is solved. Now after start-stop and then start again, it shows the last stopped time for about 1s.

**Reply by Mark-Krensel** [reply]
For anyone interested: the solution was to reset all states to 0 again in the return statement :)
